posts/ngrx-facade-pattern.md

```diff
@@ -0,0 +1,137 @@
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private store: Store<AccountListState>) {}
+
+  ngOnInit() {
+    this.accountList = this.store.select(getAccountList);
+
+    this.loadAccountList();
+  }
+
+  loadAccountList() {
+    this.store.dispatch(new LoadAccountList());
+  }
+
+  addAccount(account: string) {
+    this.store.dispatch(new AddAccount(account));
+  }
+
+  editAccount(id: string, account: string) {
+    this.store.dispatch(new EditAccount({ id, account }));
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+This implementation is very `store` dependent and makes the component more complex. By adding a facade for the component to interact with, we can simplify the access to our store and decouple the knowledge of NgRx in our component. By performing that separation of duties, if we choose to remove NgRx in the future we do not have to refactor our component.
+
+Now let's create an `AccountService` that will act as our facade for NgRx.
+``` Typescript
+
+@Injectable({
+  providedIn: 'root',
+})
+export class AccountService {
+  accountList = this.store.select(getAccountList);
+
+  constructor(private store: Store<AccountListState>) {}
+
+  loadAccountList() {
+    this.store.dispatch(new LoadAccountList());
+  }
+
+  addAccount(account: string) {
+    this.store.dispatch(new AddAccount(account));
+  }
+
+  editAccount(id: string, account: string) {
+    this.store.dispatch(new EditAccount({ id, account }));
+  }
+
+  deleteAccount(id: string) {
+    this.store.dispatch(new DeleteAccount(id));
+  }
+}
+
+```
+
+After moving everything related to the store into our service, we can then inject that service into our component.
+
+``` Typescript
+
+export class AccountListComponent implements OnInit {
+  accountList: Observable<Account[]>;
+
+  constructor(private accountService: AccountService) {}
+
+  ngOnInit() {
+    this.accountList = this.accountService.accountList;
+
+    this.accountService.loadAccountList();
+  }
+
+  addAccount(account: string) {
+    this.accountService.addAccount(account);
+  }
+
+  editAccount(id: string, account: string) {
+    this.accountService.editAccount( id, account);
+  }
+
+  deleteAccount(id: string) {
+    this.accountService.deleteAccount(id);
+  }
+}
+
+```
+After implementing the facade pattern what did we actually gain?
+
+### Pros
+- Better developer experience.
+    - Devs only interact with a service instead of NgRx stores, actions, and reducers.
+    - You can refactor your component without touching the store logic. 
+    - You can change your state management implementation without touching your component.
+- This service can be reused by other components that need to access our `accountList` instead of duplicating code.
+- Scalability.
+
+### Cons
+- Loss of indirection violates a redux pattern.
```

Comment:
Is this really "_loss of_ indirection violates a pattern" or "_indirection_ violates a pattern"? Reading the sub-bullet, I'm really not sure. Is `indirection` a Redux concept/term (in which case, could you link to some resource on it?) or your word choice (in which case, could you opt for something more familiar to the reader or otherwise make clearer?)?

Reply:
updated. I tried to make that clear in the sub-bullet